### PR TITLE
Governance MVP

### PR DESCRIPTION
**Pull request by bitparadigm**
Governance MVP: initial framework Governance BitDAO contract, docs, scripts.

### CODE REVIEW

**Comment by CjHare** [comment]
Trying to run locally I'm getting a warning due to the size 🤔 

```
Compiling 15 files with 0.8.4
Warning: Contract code size exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
  --> contracts/BitDAO.sol:22:1:
   |
22 | contract Governance is Context, ERC165, EIP712 {
   | ^ (Relevant source part starts here and spans across multiple lines).
```

**Comment by CjHare** [comment]
> I added bytecode size scripts. Its optimal to enable optimizer with 200 runs, in that case doesn't exceed for me. Many contracts (Balancer for example) securely deployed with compiler optimizations.

Though that might be the case. Didn't get to play around with the optimisation settings, but it's good to know that angle is covered 👍

File: docs/Architecture.md [hunk 1]
```diff
@@ -53,6 +53,18 @@ registerNewRole(role)
 registerNewMember(role, member)
 registerNewRole and registerNewMember will be done by accepting proposal through voting.
 
+Actions that a role can take are restricted to dao and treasury functions as these actions are within the scope of this spec and can be verified. However, note that roles can take any action since any action can be set in a proposal. 
+
+Examples of initial roles and their actions:
```
Comment (CjHare) [inline docs/Architecture.md:58]: Do we have the Governance contract init'd with any roles?

File: package.json [hunk 2]
```diff
@@ -1,24 +1,26 @@
...
+    "url": "git+https://github.com/bitdao-io/governance.git"
   },
   "keywords": [],
   "author": "",
   "license": "Apache-2.0",
   "scripts": {
-    "build": "hardhat compile",
-    "clean": "hardhat clean",
-    "format": "prettier ./**/*.ts --write",
+    "build": "npx hardhat compile",
```
Comment (CjHare) [inline package.json:17]: Out of interest do these commands not work for you without the `npx` part?

File: contracts/Governance.sol [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/governance/TimelockController.sol';
+
+// BitDAO token contract interface.
+interface IERC20Votes is IERC20 {
+    function getCurrentVotes(address account) external returns (uint256);
+}
+
+// Governance contract.
+contract Governance is Context, ERC165, EIP712 {
```
Comment (CjHare) [inline contracts/Governance.sol:22]: For ERC165 do we need an implementation of `supportsInterface` that covers all the interfaces the `Governance` implements?

File: contracts/Governance.sol [hunk 4]
```diff
@@ -0,0 +1,925 @@
...
+     */
+    receive() external payable virtual {
+        require(_executor() == address(this));
+    }
+
+    function version() public pure virtual returns (string memory) {
+        return '0.0.1';
+    }
+
+    function name() public pure virtual returns (string memory) {
```
Comment (CjHare) [inline contracts/Governance.sol:175]: `BitDAO` is the same of the ERC20 contract, perhaps something else would suit this contract better, `Governance` perhaps?

File: contracts/Governance.sol [hunk 5]
```diff
@@ -0,0 +1,925 @@
...
+        uint64 snapshot = block.number.toUint64() + votingDelay().toUint64();
+        uint64 deadline = snapshot + votingPeriod().toUint64();
+
+        proposal.voteStart.setDeadline(snapshot);
+        proposal.voteEnd.setDeadline(deadline);
+
+        return proposalId;
+    }
+
+    function registerNewRole(bytes32 role) external onlyGovernance {
```
Comment (CjHare) [inline contracts/Governance.sol:303]: How would we go about removing an existing role ...if a role that existed became redundant and was voted out by a proposal?

**Review by CjHare — COMMENTED** [review]
As a PoC it looks good. 

Although I've not been able to successfully test the contract locally, as I'm not sure how to delegate votes before a role get created, as there are no roles to begin with, only the governance contract can create roles and until I have an account with votes delegated I can't create a proposal to create a role 🤔 
_(of course, I could be overlooking something)_

File: contracts/Governance.sol [hunk 6]
```diff
@@ -0,0 +1,925 @@
...
+        bytes32[] memory actions,
+        string memory description
+    ) public virtual returns (uint256) {
+        for (uint256 i = 0; i < roles.length; ++i) {
+            require(
+                votersRoles[roles[i]][_msgSender()],
+                'Governance::propose: Proposer must have proposal roles'
+            );
+        }
+        require(
```
Comment (CjHare) [inline contracts/Governance.sol:253]: These three `require` checks all have the same message, is that right?

File: contracts/Governance.sol [hunk 7]
```diff
@@ -0,0 +1,943 @@
...
+            votersRoles[role][delegatee],
+            'Governance::hasRole: delegatee does not have the role'
+        );
+        _;
+    }
+
+    constructor(IERC20Votes token_, TimelockController timelock_) EIP712(name(), version()) {
+        token = token_;
+        _timelock = timelock_;
```
Comment (CjHare) [inline contracts/Governance.sol:159]: Doesn't `Governance` need to have the role of proposer in the `TimelockController` contract to be able to call `_timelock.scheduleBatch()` in `queue()`?
 
Where is that happening in Governance?

File: contracts/Governance.sol (hunk 5 above)
Comment (bitparadigm) [reply contracts/Governance.sol:321]: we don't for now, we just don't want on it

File: contracts/Governance.sol (hunk 4 above)
Comment (bitparadigm) [reply contracts/Governance.sol:193]: I think this still needs to be discussed, but I can integrate into governance some of the ERC20 comp func.

File: contracts/Governance.sol (hunk 7 above)
Comment (bitparadigm) [reply contracts/Governance.sol:159]: You just set it in TimelockController on its deployment.

File: contracts/Governance.sol (hunk 3 above)
Comment (bitparadigm) [reply contracts/Governance.sol:22]: No, at least not for now, its just for override

**Comment by CjHare** [comment]
While playing around with Governance contract, I ran into an error when delegating. Underfow / Overflow might warrant a require guard with appropriate message ...but with a prototype maybe not 🤷  
```
Error: VM Exception while processing transaction: reverted with panic code 0x11 (Arithmetic operation underflowed or overflowed outside of an unchecked block)
      at Governance._delegate (contracts/Governance.sol:532)
      at Governance.delegate (contracts/Governance.sol:442)
```

File: contracts/Treasury.sol [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        returns (bool);
+
+    function decreaseAllowance(address spender, uint256 subtractedValue)
+        external
+        returns (bool);
+}
+
+// Treasury contract.
+contract Treasury is Context, Ownable {
+    address public governance;
```
Comment (CjHare) [inline contracts/Treasury.sol:22]: Why extend `Ownable`, when not using the modifier it provides?

File: package.json [hunk 9]
```diff
@@ -1,24 +1,26 @@
...
+    "url": "git+https://github.com/bitdao-io/governance.git"
   },
   "keywords": [],
   "author": "",
   "license": "Apache-2.0",
   "scripts": {
-    "build": "hardhat compile",
-    "clean": "hardhat clean",
-    "format": "prettier ./**/*.ts --write",
+    "compile": "npx hardhat compile",
```
Comment (CjHare) [inline package.json:17]: I'd recommend against changing the build command to compile, reasoning:
- consistency with TypeScript/JavaScript projects; a build target is the common one used for building the project.
- compile is an inaccurate term; While Solidity is compiled, TypeScript is transpiled, while similar is not equivalent, specificity can be valuable.
- build target is supported as default script by Npm, while compile is not https://docs.npmjs.com/cli/v6/commands/npm-build

A solution here can be to have both `build` and `compile` targets 🙂

**Comment by CjHare** [comment]
This currently being a PoC, you're welcome to squash & merge the PR at anytime, as any comments I'm making are only suggestions.

I'd usually recommend merging the open PR and creating a new branch & Pr for any substantial changes or additions widening the scope. It makes it far easier & quicker to review smaller pieces / changes and also can provide a more navigable git history.

**Review by shahankhatch — APPROVED** [review]
lgtm. This should make it into `main` so that others can work on it at the same time.

File: contracts/Treasury.sol [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    mapping(address => uint256) public balances;
+    bytes32 public name = 'TREASURY';
+
+    event Received(address from, address asset, uint256 amount);
+    event Sent(address to, address asset, uint256 amount);
+    event IncreasedAllowance(address spender, address asset, uint256 amount);
+    event DecreasedAllowance(address spender, address asset, uint256 amount);
+
+    modifier onlyGovernance() {
```
Comment (CjHare) [inline contracts/Treasury.sol:32]: This modifier is not being used, is it needed? ...could the `governance` field also be removed?

**Review by CjHare — APPROVED** [review]
Time to merge this PR into main 👍

**Review by shahankhatch — APPROVED** [review]

**Pull request merged**